apps/skyspy/skyspy.app.js

```diff
@@ -0,0 +1,261 @@
+    .drawString("alt "+alt, 0, 110)
+    .drawString("balt " + cur_altitude, 0, 130)
+    .drawString(step, 0, 150);
+    step++;
+    if (step == 10) {
+      qalt = max_dalt - min_dalt;
+      resetAlt();
+    }
+  }
+
+  if (debug > 0)
+    print(fix);
+  setTimeout(updateGps, 1000);
+}
+
+function radLine(a1, d1, a2, d2) {
+  g.drawLine(radX(a1, d1), radY(a1, d1), radX(a2, d2), radY(a2, d2));
+}
+function radCircle(d) {
+  let step = 0.05;
+  for (let i=0; i<1; i+=0.05) {
+    radLine(i-step, d, i, d);
+  }
+  //g.flip();
+}
+function drawGrid() {
+  g.setColor(0,0,0);
+  radLine(0, 1, 0.5, 1);
+  radLine(0.25, 1, 0.75, 1);
+  radCircle(0.5);
+  radCircle(1.0);
+}
+function drawSat(s) {
+  let a = s.azi / 360;
+  let e = ((90 - s.ele) / 90);
+  let x = radX(a, e);
+  let y = radY(a, e);
+
+  if (s.snr == "")
+    g.setColor(1, 0.25, 0.25);
+  else {
+    let snr = 1*s.snr;
+    g.setColor(0, 0, 0);
+    sats_receiving ++;
+  }
+  g.drawString(s.id, x, y);
+
+}
+
+// Should correspond to view from below.
+// https://in-the-sky.org//satmap_radar.php?year=2023&month=10&day=24&skin=1
+function drawSats(sats) {
+  sats_receiving = 0;
+  g.reset().setFont("Vector", 20)
+    .setColor(1,1,1)
+    .fillRect(0, 30, 176, 176);
+
+  drawGrid();
+  for (var s of sats) {
+    if (debug > 1)
+      print(s.ele, s.azi, s.snr);
+    drawSat(s);
+  }
+  if (fix && fix.fix && fix.lat) {
+    g.setColor(0, 0, 0);
+    g.drawString(fix.satellites + "/" + fix.hdop, 10, 150);
+  }
+}
+
+var sats = [];
+var snum = 0;
+var sats_receiving = 0;
+
+function parseRaw(msg, lost) {
+  if (lost)
+    print("## data lost");
+  let s = msg.split(",");
+  if (s[0] != "$GPGSV")
+    return;
+  //print("Message", s[2], s[1]);
+  if (debug > 0)
+    print(msg);
+
+  if (s[2] == "1") {
+    snum = 0;
+    sats = [];
+  }
+
+  let view = 1 * s[3];
+
+  // s[3] -- sats in view.
+  // id, ele, azi, snr
+  if (debug > 0)
+    print("in view:", view);
+  let i = 4;
+  let k = 4;
+  if (view - snum < k)
+    k = view - snum;
+  for (let j=0; j<k; j++) {
+    let sat = {};
+    sat.id = s[i++];
+    sat.ele = 1*s[i++];
+    sat.azi = 1*s[i++];
+    sat.snr = s[i++];
+    if (debug > 0)
+      print("  ", sat);
+    sats[snum++] = sat;
+  }
+  if (debug > 1)
+    print("Checksum:", s[i]);
+  if (s[1] == s[2]) {
+    print("Complete...");
+    //print(sats);
+    if (display == 0)
+      drawSats(sats);
+  }
+}
+
+function stopGps() {
+  cancel_gps=true;
+  Bangle.setGPSPower(0, "skyspy");
+}
+
+function markGps() {
+  cancel_gps = false;
+  Bangle.setGPSPower(1, "skyspy");
+  Bangle.on('GPS-raw', parseRaw);
+  gps_start = getTime();
+  updateGps();
+}
+
+function onSwipe(dir) {
+  display = display + 1;
+  if (display == 3)
+    display = 0;
+}
+
+Bangle.setUI({
+  mode : "custom",
+  back : function() {}, // optional - add a 'back' icon in top-left widget area and call this function when it is pressed , also call it when the hardware button is clicked (does not override btn if defined)
```

Comment:
Not sure about this part: `back : function() {}` doesn't this cause the fastload of the cloak if 'fastload' is installed?

The whole code doesn't seem to be compatible with fastload so I would remove this line

Reply:
> Not sure about this part: back : function() {} doesn't this cause the fastload[1] of the cloak if 'fastload' is installed?

I'm pretty sure this is not the case (please enlighten me if I'm wrong!). It's the presence of the `remove` key that allows fastloading to trigger (*https://www.espruino.com/Bangle.js+Fast+Load#1-bangle-setui-remove-*).

The `back` key is used to :
```
back : function() {}, // optional - add a 'back' icon in top-left widget area and call this function when it is pressed , also call it when the hardware button is clicked (does not override btn if defined)
```
*https://www.espruino.com/ReferenceBANGLEJS2#l_Bangle_setUI*